hail/python/hail/matrixtable.py

```diff
@@ -2807,25 +2807,205 @@ def write(
         Env.backend().execute(ir.MatrixWrite(self._mir, writer))
 
     class _Show:
-        def __init__(self, table, n_rows, actual_n_cols, displayed_n_cols, width, truncate, types):
-            self.table_show = table._show(n_rows, width, truncate, types)
-            self.actual_n_cols = actual_n_cols
-            self.displayed_n_cols = displayed_n_cols
+        def __init__(
+            self,
+            matrix_table: 'MatrixTable',
+            n_rows: Optional[int] = None,
+            n_cols: Optional[int] = None,
+            include_row_fields: bool = False,
+            width: Optional[int] = None,
+            truncate_limit: Optional[int] = None,
+            should_show_types: bool = True,
+        ):
+            self.matrix_table = matrix_table
+            self.n_rows = n_rows
+            self.n_cols = n_cols
+            self.include_row_fields = include_row_fields
+            self.width = width
+            self.truncate_limit = truncate_limit
+            self.should_show_types = should_show_types
 
         def __str__(self):
-            s = self.table_show.__str__()
-            if self.displayed_n_cols != self.actual_n_cols:
-                s += f"showing the first { self.displayed_n_cols } of { self.actual_n_cols } columns"
-            return s
+            matrix_table = self.matrix_table
+            n_rows = self.n_rows
+            n_cols = self.n_cols
+            include_row_fields = self.include_row_fields
+            width = self.width
+            truncate_limit = self.truncate_limit
+            should_show_types = self.should_show_types
+
+            def truncate(string: str):
+                return string[: truncate_limit - 3] + "..." if len(string) > truncate_limit else string
+
+            table = matrix_table.head(n_rows, n_cols).localize_entries('entries', 'cols')
+            table_key_dtype = table.key.dtype
+            if len(table.key) > 0:
+                # order_by unkeys the table
+                table = table.order_by(*table.key)
+
+            row_dtype = table.row.dtype
+            rows, _ = table._take_n(n_rows)
+            row_fields = list(row_dtype) if include_row_fields else list(table_key_dtype)
+            row_fields = [row_field for row_field in row_fields if row_field != 'entries']
+            truncated_row_fields = [truncate(field) for field in row_fields]
+            truncated_rows = [[truncate(str(row[field])) for field in row_fields] for row in rows]
```

Comment:
Instead of str I think we should use `repr`. In particular, adversarial inputs produce break the output:
```
In [6]: hl.utils.range_matrix_table(10,1000).annotate_entries(a='a\n').show(5, 2)
+---------+-----+-----+
| row_idx | 0.a | 1.a |
+---------+-----+-----+
|   int32 | str | str |
+---------+-----+-----+
|       0 | a
  | a
  |
|       1 | a
  | a
  |
|       2 | a
  | a
  |
|       3 | a
  | a
  |
|       4 | a
  | a
  |
+---------+-----+-----+
showing top 5 rows
showing the first 2 of 1000 columns
```